examples/opengl/cube/source/app.d

```
@@ -0,0 +1,533 @@
+			Camera* camera = world.query!Camera.each.front[1]; // only one camera
+			vec4 clip = vec4(mouseX, mouseY, -1f, 1f);
+
+			with (camera)
+			{
+				// dlib doesn't allow Matrix * Vector, so we need to transpose
+				vec4 eye;
+				with (clip * projection.inverse.transposed) eye = vec4(x, y, -1f, 0f);
+				vec3 ray = (eye * view.inverse.transposed).xyz.normalized;
+
+				// for some reason these are nothrow
+				Plane plane;
+				plane.fromPointAndNormal(vec3(0, 0, 0), vec3(0, 0, 1f)).assumeWontThrow;
+				plane.intersectsLineSegment(position, position + ray.normalized, ray).assumeWontThrow;
+
+				return ray;
+			}
+		};
+
+
+		if (mods & GLFW_MOD_CONTROL && button == GLFW_MOUSE_BUTTON_LEFT)
+		{
+			// spawn a cube with pivot at '0, 0, 0' translated to 'x, y, z'
+			Entity entity;
+			with (screenToWorld(vec2(mouseX, mouseY))) entity = world.entity
+					.emplace!Transform(vec3(0, 0, 0), vec3(1, 1, 1), vec3(0, 0, 0))
+					.emplace!Model(vec3(x, y, z), vec3(1, 1, 1), vec3(0, 0, 0))
+					.add!Graphics
+					.add!Input;
+			version(VECS_EXAMPLE_DEBUG) tracef("[VECS] entity '%d' created", entity).assumeWontThrow;
+		}
+		else if (button == GLFW_MOUSE_BUTTON_LEFT)
+		{
+			// spawn a cube with pivot at 'x, y, z'
+			Entity entity;
+			with (screenToWorld(vec2(mouseX, mouseY))) entity = world.entity
+					.emplace!Transform(vec3(x, y, z), vec3(1, 1, 1), vec3(0, 0, 0))
+					.emplace!Model(vec3(0, 0, 0), vec3(1, 1, 1), vec3(0, 0, 0))
+					.add!Graphics
+					.add!Input;
+			version(VECS_EXAMPLE_DEBUG) tracef("[VECS] entity '%d' created", entity).assumeWontThrow;
+		}
+		else if (button == GLFW_MOUSE_BUTTON_RIGHT)
+		{
+			// destroy the most recent spawned cube
+			foreach_reverse(entity; world.query!Transform)
+			{
+				world.destroyEntity(entity);
+				version(VECS_EXAMPLE_DEBUG) tracef("[VECS] entity '%d' destroyed", entity).assumeWontThrow;
+				break;
+			}
+		}
+	}
+}
+
+void keyCallback(GLFWwindow* window, int key, int, int action, int)
+{
+	switch (action)
```

If those are the only actions possible in this case, this should be a `final switch`, otherwise it is silently skipping an unknown key action.

---

There is one more action `GLFW_REPEAT` that I want to ignore. This action is fired when a key stays pressed. But I don't care about it because I'm using an input manager. And that information is retrieved automatically with only `GLFW_PRESS` and `GLFW_RELEASE`.